File: android/src/main/java/com/memotalk/sdk/MemoTalkSDK.kt
package com.memotalk.sdk

import android.content.Context
import java.io.File

object MemoTalkSDK {

    private var instance: MemoTalkSDKImpl? = null

    fun initialize(context: Context, config: SDKConfig): Result<Unit, PlaudError> {
        return try {
            instance = MemoTalkSDKImpl(context, config)
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(PlaudError(0, "SDK initialization failed: ${e.message}"))
        }
    }

    fun getInstance(): MemoTalkSDKImpl? = instance

    fun getDeviceManager(): DeviceManager? = instance?.deviceManager
    fun getRecordingManager(): RecordingManager? = instance?.recordingManager
    fun getFileManager(): FileManager? = instance?.fileManager
    fun getOTAManager(): OTAManager? = instance?.otaManager
    fun getTranscriptionManager(): TranscriptionManager? = instance?.transcriptionManager

    fun release() {
        instance?.release()
        instance = null
    }
}

class MemoTalkSDKImpl(
    private val context: Context,
    private val config: SDKConfig
) {

    val deviceManager: DeviceManager
    val recordingManager: RecordingManager
    val fileManager: FileManager
    val otaManager: OTAManager
    val transcriptionManager: TranscriptionManager

    private val bleClient: BleClient

    init {
        bleClient = BleClient(context)
        deviceManager = DeviceManager(context)
        recordingManager = RecordingManager(bleClient)
        fileManager = FileManager(bleClient, context.cacheDir)
        otaManager = OTAManager(bleClient)
        transcriptionManager = TranscriptionManager(bleClient)
    }

    fun release() {
        bleClient.close()
    }
}

File: ios/MemoTalkSDK.swift
import Foundation

public class MemoTalkSDK {

    private static var instance: MemoTalkSDKImpl?

    public static func initialize(config: SDKConfig) -> Result<Void, PlaudError> {
        do {
            instance = MemoTalkSDKImpl(config: config)
            return .success(())
        } catch {
            return .failure(PlaudError(code: 0, message: "SDK initialization failed: \(error.localizedDescription)"))
        }
    }

    public static func getInstance() -> MemoTalkSDKImpl? {
        return instance
    }

    public static func getDeviceManager() -> DeviceManager? {
        return instance?.deviceManager
    }

    public static func getRecordingManager() -> RecordingManager? {
        return instance?.recordingManager
    }

    public static func getFileManager() -> FileManager? {
        return instance?.fileManager
    }

    public static func getOTAManager() -> OTAManager? {
        return instance?.otaManager
    }

    public static func getTranscriptionManager() -> TranscriptionManager? {
        return instance?.transcriptionManager
    }

    public static func release() {
        instance?.release()
        instance = nil
    }
}

public class MemoTalkSDKImpl {

    public let deviceManager: DeviceManager
    public let recordingManager: RecordingManager
    public let fileManager: FileManager
    public let otaManager: OTAManager
    public let transcriptionManager: TranscriptionManager

    private let bleClient: BleClient
    private let config: SDKConfig

    public init(config: SDKConfig) {
        self.config = config
        bleClient = BleClient()
        deviceManager = DeviceManager()
        recordingManager = RecordingManager(bleClient: bleClient)
        fileManager = FileManager(bleClient: bleClient)
        otaManager = OTAManager(bleClient: bleClient)
        transcriptionManager = TranscriptionManager(bleClient: bleClient)
    }

    public func release() {
        bleClient.disconnect()
    }
}
